research/2d-decoding/rectangle_detector.py

```diff
@@ -0,0 +1,220 @@
+# Backend code to find the rectangles
+
+from PIL import Image
+import matplotlib.pyplot as plt
+import cv2
+
+LINE_PIXEL_TOLERANCE = 1
+
+def mins(items, key=lambda a: a):
+    items = list(items)
+    minV = key(items[0])
+    minItems = [items[0]]
+    for i in range(1, len(items)):
+        val = key(items[i])
+        if val == minV:
+            minItems.append(items[i])
+        elif val < minV:
+            minV = val
+            minItems = [items[i]]
+    return minItems
+
+def maxs(items, key=lambda a: a):
+    items = list(items)
+    maxV = key(items[0])
+    maxItems = [items[0]]
+    for i in range(1, len(items)):
+        val = key(items[i])
+        if val == maxV:
+            maxItems.append(items[i])
+        elif val > maxV:
+            maxV = val
+            maxItems = [items[i]]
+    return maxItems
+
+def readImage(filename):
+    # Clean up the image to get a nice black and white image
+    img = cv2.imread(filename)
+    # Turn it into a BW image
+    img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+    # Blur filter to remove minor noise
+    blur = cv2.GaussianBlur(img,(3,3),0)
+    
+    # Threshold the image adaptively: this is convert to binary image in small chunks, choosing intermediate threshold values based on local statistics (avoids very white parts of the immage destroying everything else)
+    thresholdedImage = cv2.adaptiveThreshold(blur,255,cv2.ADAPTIVE_THRESH_MEAN_C,cv2.THRESH_BINARY,301,2) # 301 (the block size) has been arbitrarily chosen based off, 2 is the constant to subtract from the mean, not sure what it changes
```

Comment:
"has been arbitrarily chosen based off" based on what?

Reply:
Thanks for brining this up! Chaning it to a parametric 10% in commit 948abd5 Actually significantly improves scanning!

Before: 
![image](https://user-images.githubusercontent.com/54122151/197659969-7051c24a-df6f-4a54-ae1b-f0d5e7839c43.png)

After:
![image](https://user-images.githubusercontent.com/54122151/197659699-aadfaaa6-82af-425d-bbed-d307751df8d8.png)